core/include/scipp/core/multi_index.h

```diff
@@ -30,276 +47,332 @@ auto get_slice_dim(const T &param, const Ts &... params) {
+    scipp::index *const shape_buffer = strides_buffer + strides_buffer_size;
+    m_ndim = detail::flatten_dims(strides_buffer, shape_buffer, iter_dims, 0,
+                                  strides...);
+    m_inner_ndim = m_ndim;
+
+    m_buffer =
+        std::make_unique<scipp::index[]>(detail::get_buffer_size<N>(m_ndim));
+    std::copy(strides_buffer, strides_buffer + N * m_ndim, &stride(0, 0));
+    std::copy(shape_buffer, shape_buffer + m_ndim, shape_it());
   }
 
+  /// Determine from arguments if binned.
   template <class... Params>
   explicit MultiIndex(const ElementArrayViewParams &param,
-                      const Params &... params) {
-    init(param, params...);
-  }
-
-  template <class Param> void validate_bin_indices(const Param &) {}
+                      const Params &... params)
+      : MultiIndex{
+            (!param.bucketParams() && (!params.bucketParams() && ...))
+                ? MultiIndex(param.dims(), param.strides(), params.strides()...)
+                : MultiIndex{binned_tag{},
+                             detail::get_nested_dims(param, params...),
+                             param.dims(), param, params...}} {}
 
-  /// Check that corresponding bins have matching sizes.
-  template <class Param0, class Param1, class... Params>
-  void validate_bin_indices(const Param0 &param0, const Param1 &param1,
-                            const Params &... params) {
-    if (param0.bucketParams() && param1.bucketParams())
-      validate_bin_indices_impl(param0, param1);
-    if (param0.bucketParams())
-      validate_bin_indices(param0, params...);
-    else
-      validate_bin_indices(param1, params...);
-  }
+private:
+  /// Use to disambiguate between constructors.
+  struct binned_tag {};
 
-  template <class... Params> void init(const Params &... params) {
-    const auto iterDims = std::array<Dimensions, N>{params.dims()...}[0];
-    if ((!params.bucketParams() && ...)) {
-      *this = MultiIndex(iterDims, params.strides()...);
-      return;
-    }
-    validate_bin_indices(params...);
-    const auto nestedDims = get_nested_dims(params.bucketParams()...);
-    const Dim sliceDim = get_slice_dim(params.bucketParams()...);
-    m_ndim_nested = nestedDims.ndim();
-    m_ndim = iterDims.ndim() + m_ndim_nested;
-    m_nested_stride = nestedDims.offset(sliceDim);
-    m_nested_dim_index = m_ndim_nested - nestedDims.index(sliceDim) - 1;
-    m_bin = std::array{BinIterator(params)...};
-    scipp::index dim = iterDims.ndim() - 1 + nestedDims.ndim();
-    m_end_sentinel = iterDims.volume();
-    if (m_end_sentinel == 0) {
-      return; // operands are empty, leave everything below default initialized
-    }
-    for (const auto size : iterDims.shape()) {
-      m_shape[dim--] = size;
-    }
-    for (const auto size : nestedDims.shape()) {
-      m_shape[dim--] = size;
+  /// Construct with bins.
+  template <class... Params>
+  explicit MultiIndex(binned_tag, const Dimensions &inner_dims,
```

Comment:
Kind of unrelated to this PR, but I wonder if we could improve compile times / binary sizes by moving this to the source and making explicit instantiations?

Reply:
Maybe. But which functions would you want to move? I think the core functions like `increment_outer` should be inline.

Reply:
I meant only the constructor, which is particularely heavy.